File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {

        vector<vector<int>>res;
        sort(nums.begin(),nums.end());
        if(nums.size()<3) return {};
 

        for(int i=0;i<nums.size()-2;i++)
        {
           
        if (i > 0 && nums[i] == nums[i - 1]) continue;

        // pruning: if anchor > 0, no further triplets can sum to 0
        if (nums[i] > 0) break;
            
            int target=-nums[i];

            int l=i+1;
            int r=(int)nums.size()-1;
        
           

            while(l<r)
            {
                if(nums[l]+nums[r]<target) l++;
                else if(nums[l]+nums[r]>target) r--;

                else{
                    res.push_back({nums[i],nums[l],nums[r]});
                    int left=nums[l];
                    int right=nums[r];

                    while(l<r && nums[l]==left) l++;
                    while(l<r && nums[r]==right) r--;



                }

            }

            
        }
        return res;

        

        
    }
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        # basically in this question i have to take a number and now i have to find it's triplets in the
        # remaining subspace using two pointers


        nums.sort()       #Now this will bring the duplicates together and the search space would be optimal

        ans=[]

        for i,n in enumerate(nums):

            
            l=i+1
            r=len(nums)-1
            if i>0 and nums[i-1]==nums[i]:
                continue
            
            while l<r:
                total=n+nums[l]+nums[r]

                if total>0:
                    r-=1
                elif total<0:
                    l+=1
                elif total==0:
                    ans.append([n,nums[l],nums[r]])
                    l+=1
                    r-=1
                    while l<r and nums[l]==nums[l-1]:
                        l+=1
                    while l<r and nums[r]==nums[r+1]:
                        r-=1
                
        return ans


        